File: mil/sdk-swift/Sources/MilSDK/Protocol.swift
import Foundation

/// MIL protocol types + wire codecs (design §2.3/§7.4), Borsh-compatible with
/// the Rust `misaka_mil_channel::wire` + `misaka_mil_core::{job, receipt}`.
public enum MilProtocol {
    public static let version: UInt16 = 1
    public static let ftClient: UInt8 = 0x01
    public static let ftServer: UInt8 = 0x02
}

public enum Tier: UInt8 { case tee = 0, open = 1 }

public struct SamplingParams {
    public var temperatureMilli: UInt16
    public var topPMilli: UInt16
    public var seed: UInt64?
    public init(temperatureMilli: UInt16 = 0, topPMilli: UInt16 = 1000, seed: UInt64? = nil) {
        self.temperatureMilli = temperatureMilli
        self.topPMilli = topPMilli
        self.seed = seed
    }
    public static let greedy = SamplingParams()
}

public struct SlaParams {
    public var ttfbMs: UInt32
    public var minTps: UInt32
    public init(ttfbMs: UInt32, minTps: UInt32) { self.ttfbMs = ttfbMs; self.minTps = minTps }
}

public struct JobSpec {
    public var version: UInt16 = MilProtocol.version
    public var modelId: [UInt8]      // 64
    public var profileId: [UInt8]?   // 64
    public var tier: Tier
    public var maxTokens: UInt32
    public var sampling: SamplingParams
    public var sla: SlaParams
    public var priceCapSompi: UInt64
    public var cmReq: [UInt8]        // 64

    public func encodeAsClientMsg() -> [UInt8] {
        var w = BorshWriter()
        w.u8(1)  // ClientMsg::Job
        w.u16(version); w.fixed(modelId)
        w.option(profileId != nil) { $0.fixed(profileId!) }
        w.u8(tier.rawValue)
        w.u32(maxTokens)
        w.u16(sampling.temperatureMilli); w.u16(sampling.topPMilli)
        w.option(sampling.seed != nil) { $0.u64(sampling.seed!) }
        w.u32(sla.ttfbMs); w.u32(sla.minTps)
        w.u64(priceCapSompi)
        w.fixed(cmReq)
        return w.bytes
    }
}

public struct ReceiptBody {
    public var version: UInt16
    public var sessionId: [UInt8]  // 64
    public var counter: UInt64
    public var cumTokensIn: UInt64
    public var cumTokensOut: UInt64
    public var timestampMs: UInt64
    public var cmResp: [UInt8]     // 64
    public var isFinal: Bool

    /// The canonical 163-byte receipt signing transcript (LE ints).
    public func signingMessage() -> [UInt8] {
        var w = BorshWriter()
        w.u16(version); w.fixed(sessionId)
        w.u64(counter); w.u64(cumTokensIn); w.u64(cumTokensOut); w.u64(timestampMs)
        w.fixed(cmResp); w.u8(isFinal ? 1 : 0)
        return w.bytes
    }
}

public struct SignedReceipt {
    public var body: ReceiptBody
    public var signature: [UInt8]  // 4627
    public var providerPk: [UInt8] // 2592
}

// Wire messages.

public struct ServerHello {
    public var version: UInt16
    public var attestation: [UInt8]
    public var pkKem: [UInt8]
    public var pkReceipt: [UInt8]

    public static func decode(_ buf: [UInt8]) -> ServerHello {
        var r = BorshReader(buf)
        return ServerHello(version: r.u16(), attestation: r.vecU8(), pkKem: r.vecU8(), pkReceipt: r.vecU8())
    }
}

public enum ClientMsg {
    public static func prompt(_ p: [UInt8]) -> [UInt8] {
        var w = BorshWriter(); w.u8(0); w.vecU8(p); return w.bytes
    }
    public static let cancel: [UInt8] = { var w = BorshWriter(); w.u8(2); return w.bytes }()
}

public enum ServerMsg {
    case chunk(text: [UInt8], tokenCount: UInt32)
    case receipt(SignedReceipt)
    case done(totalTokensOut: UInt64)
    case error(String)

    public static func decode(_ buf: [UInt8]) -> ServerMsg {
        var r = BorshReader(buf)
        switch r.u8() {
        case 0: return .chunk(text: r.vecU8(), tokenCount: r.u32())
        case 1:
            let body = ReceiptBody(
                version: r.u16(), sessionId: r.fixed(64), counter: r.u64(),
                cumTokensIn: r.u64(), cumTokensOut: r.u64(), timestampMs: r.u64(),
                cmResp: r.fixed(64), isFinal: r.bool())
            return .receipt(SignedReceipt(body: body, signature: r.vecU8(), providerPk: r.vecU8()))
        case 2: return .done(totalTokensOut: r.u64())
        case 3: return .error(r.string())
        default: return .error("unknown ServerMsg tag")
        }
    }
}

public func encodeClientHello(nonceReq: [UInt8]) -> [UInt8] {
    var w = BorshWriter(); w.u16(MilProtocol.version); w.fixed(nonceReq); return w.bytes
}
public func encodeClientKem(kemCt: [UInt8]) -> [UInt8] {
    var w = BorshWriter(); w.vecU8(kemCt); return w.bytes
}
public func encodeFrame(frameType: UInt8, seq: UInt64, ciphertext: [UInt8]) -> [UInt8] {
    var w = BorshWriter(); w.u8(frameType); w.u64(seq); w.vecU8(ciphertext); return w.bytes
}
public func decodeFrame(_ buf: [UInt8]) -> (frameType: UInt8, seq: UInt64, ciphertext: [UInt8]) {
    var r = BorshReader(buf); return (r.u8(), r.u64(), r.vecU8())
}

File: mil/sdk-ts/src/protocol.ts
// MIL wire message codecs (design §2.3/§7.4). Borsh-compatible with the Rust
// `misaka_mil_channel::wire` + `misaka_mil_core::{job, receipt}`.

import { BorshReader, BorshWriter } from "./borsh.ts";
import type { ReceiptBody, SignedReceipt } from "./receipt.ts";

export const MIL_PROTOCOL_VERSION = 1;
export const FT_CLIENT = 0x01;
export const FT_SERVER = 0x02;

export const Tier = { Tee: 0, Open: 1 } as const;
export type Tier = (typeof Tier)[keyof typeof Tier];

export interface SamplingParams {
  temperatureMilli: number;
  topPMilli: number;
  seed: bigint | null;
}

export interface SlaParams {
  ttfbMs: number;
  minTps: number;
}

export interface JobSpec {
  version: number;
  modelId: Uint8Array; // 64
  profileId: Uint8Array | null; // 64
  tier: Tier;
  maxTokens: number;
  sampling: SamplingParams;
  sla: SlaParams;
  priceCapSompi: bigint;
  cmReq: Uint8Array; // 64
}

// --- handshake ---

export function encodeClientHello(nonceReq: Uint8Array): Uint8Array {
  return new BorshWriter().u16(MIL_PROTOCOL_VERSION).fixed(nonceReq).finish();
}

export interface ServerHello {
  version: number;
  attestation: Uint8Array;
  pkKem: Uint8Array;
  pkReceipt: Uint8Array;
}

export function decodeServerHello(buf: Uint8Array): ServerHello {
  const r = new BorshReader(buf);
  return { version: r.u16(), attestation: r.bytes(), pkKem: r.bytes(), pkReceipt: r.bytes() };
}

export function encodeClientKem(kemCt: Uint8Array): Uint8Array {
  return new BorshWriter().bytes(kemCt).finish();
}

// --- encrypted frame ---

export interface EncryptedFrame {
  frameType: number;
  seq: bigint;
  ciphertext: Uint8Array;
}

export function encodeFrame(f: EncryptedFrame): Uint8Array {
  return new BorshWriter().u8(f.frameType).u64(f.seq).bytes(f.ciphertext).finish();
}

export function decodeFrame(buf: Uint8Array): EncryptedFrame {
  const r = new BorshReader(buf);
  return { frameType: r.u8(), seq: r.u64(), ciphertext: r.bytes() };
}

// --- sealed application messages ---

// ClientMsg: 0=Prompt(Vec<u8>), 1=Job(JobSpec), 2=Cancel
export function encodeClientPrompt(prompt: Uint8Array): Uint8Array {
  return new BorshWriter().u8(0).bytes(prompt).finish();
}

export function encodeJobSpec(job: JobSpec): Uint8Array {
  const w = new BorshWriter().u8(1); // ClientMsg::Job tag
  writeJobSpec(w, job);
  return w.finish();
}

function writeJobSpec(w: BorshWriter, j: JobSpec): void {
  w.u16(j.version).fixed(j.modelId);
  w.option(j.profileId, (ww, id) => ww.fixed(id));
  w.u8(j.tier);
  w.u32(j.maxTokens);
  w.u16(j.sampling.temperatureMilli).u16(j.sampling.topPMilli);
  w.option(j.sampling.seed, (ww, s) => ww.u64(s));
  w.u32(j.sla.ttfbMs).u32(j.sla.minTps);
  w.u64(j.priceCapSompi);
  w.fixed(j.cmReq);
}

export function encodeClientCancel(): Uint8Array {
  return new BorshWriter().u8(2).finish();
}

// ServerMsg: 0=Chunk{text,token_count}, 1=Receipt(SignedReceipt), 2=Done{total}, 3=Error(String)
export type ServerMsg =
  | { kind: "chunk"; text: Uint8Array; tokenCount: number }
  | { kind: "receipt"; receipt: SignedReceipt }
  | { kind: "done"; totalTokensOut: bigint }
  | { kind: "error"; message: string };

export function decodeServerMsg(buf: Uint8Array): ServerMsg {
  const r = new BorshReader(buf);
  const tag = r.u8();
  switch (tag) {
    case 0:
      return { kind: "chunk", text: r.bytes(), tokenCount: r.u32() };
    case 1:
      return { kind: "receipt", receipt: readSignedReceipt(r) };
    case 2:
      return { kind: "done", totalTokensOut: r.u64() };
    case 3:
      return { kind: "error", message: r.string() };
    default:
      throw new Error(`MIL: unknown ServerMsg tag ${tag}`);
  }
}

function readReceiptBody(r: BorshReader): ReceiptBody {
  return {
    version: r.u16(),
    sessionId: r.fixed(64),
    counter: r.u64(),
    cumTokensIn: r.u64(),
    cumTokensOut: r.u64(),
    timestampMs: r.u64(),
    cmResp: r.fixed(64),
    isFinal: r.bool(),
  };
}

export function readSignedReceipt(r: BorshReader): SignedReceipt {
  return { body: readReceiptBody(r), signature: r.bytes(), providerPk: r.bytes() };
}

export function encodeSignedReceipt(rec: SignedReceipt): Uint8Array {
  const w = new BorshWriter();
  const b = rec.body;
  w.u16(b.version).fixed(b.sessionId).u64(b.counter).u64(b.cumTokensIn).u64(b.cumTokensOut).u64(b.timestampMs).fixed(b.cmResp).bool(b.isFinal);
  w.bytes(rec.signature).bytes(rec.providerPk);
  return w.finish();
}
